### ISSUE TITLE
Error "undefined-variable" for `TYPE_CHECKING` code

### ISSUE THREAD

**Issue by Flamefire** [pylint-dev/pylint#10778]
### Bug description

Using the common pattern of importing types only in `TYPE_CHECKING` mode causes an E0602:

```python
# pylint: disable=missing-docstring,

TYPE_CHECKING = False
if TYPE_CHECKING:
    from typing import Any

def conv(o: Any) -> str:
    return str(o)
```

### Command used

```shell
pylint /tmp/test.py
```

### Pylint output

```python
************* Module test
/tmp/test.py:7:12: E0602: Undefined variable 'Any' (undefined-variable)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 3.3.9
astroid 3.3.8
Python 3.9.18 (main, Sep 17 2024, 16:28:06)
```

**Comment by Pierre-Sassoulas** [comment]
This seem to be working as intended because what's imported from typing is ``IO`` and not ``Any`` so ``Any`` is not actually defined ?

**Comment by Flamefire** [comment]
Must have mispasted when reducing the actual code Edited.

On this machine with pylint 4.0.4 I get instead:
> E0601: Using variable 'Any' before assignment (used-before-assignment)

**Comment by Pierre-Sassoulas** [comment]
You need to either use ``from __future__ import annotations`` at the beginning of the file or ``def conv(o: "Any") -> str:`` instead of ``def conv(o: Any) -> str:``, or add the ``--py-version=3.14`` option when launching pylint. (pylint is still compatible with python 3.10, I'm not exactly sure when not using string on typing conditionally imported with TYPING_CHECK guard became valid python, but apparently 3.14).

**Comment by Flamefire** [comment]
> You need to either use `from __future__ import annotations` at the beginning of the file or `def conv(o: "Any") -> str:` instead of `def conv(o: Any) -> str:`, or add the `--py-version=3.14` option when launching pylint. (pylint is still compatible with python 3.10, I'm not exactly sure when not using string on typing conditionally imported with TYPING_CHECK guard became valid python, but apparently 3.14).

Thanks didn't know that.   
Maybe that would be worth a check message? I.e. usage of `if TYPING_CHECK` without the future import on a lower Python version than 3.14`. Could be a warning
